main.py

```diff
@@ -1,29 +1,89 @@
 """
 This is an Add-On for DocumentCloud which wraps the tabula-py library
 https://github.com/chezou/tabula-py
-
-It attempts to convert tables from PDFs to CSVs
+It allows you to export tables from a set of PDFs into CSVs and
+download the resulting CSVs as a zip file.
 """
 
+import os
+import sys
+import zipfile
+import subprocess
+from urllib.parse import urlparse
+
+import requests
 import tabula
+
 from documentcloud.addon import AddOn
+import lootdl
+from lootdl import DROPBOX_URL, GDRIVE_URL, MEDIAFIRE_URL, WETRANSFER_URL
 
 
 class Tabula(AddOn):
     """A tabula Add-On for DocumentCloud"""
 
-    def main(self):
+    def fetch_files(self, url):
+        """Fetch the files from either a cloud share link or any public URL"""
+        self.set_message("Downloading the files...")
 
-        document = self.client.documents.get(self.documents[0])
-        with open(f"{document.slug}.pdf", "wb") as pdf_file:
-            pdf_file.write(document.pdf)
+        os.makedirs(os.path.dirname("./out/"), exist_ok=True)
+        cloud_urls = [DROPBOX_URL, GDRIVE_URL, MEDIAFIRE_URL, WETRANSFER_URL]
+        if any(cloud_url in url for cloud_url in cloud_urls):
+            # surpress output during lootdl download to avoid leaking
+            # private information
+            stdout = sys.stdout
+            sys.stdout = open(os.devnull, "w")
+            lootdl.grab(url, "./out/")
+            # restore stdout
+            sys.stdout = stdout
+            rename_file = "cd out; mv * template.json"
+            subprocess.call(rename_file, shell=True)
 
-        tabula.convert_into(
-            f"{document.slug}.pdf", f"{document.slug}.csv", output_format="csv", pages="all"
-        )
+        else:
+            parsed_url = urlparse(url)
+            basename = os.path.basename(parsed_url.path)
+            title, ext = os.path.splitext(basename)
+            if not title:
+                title = "template"
+            if not ext:
+                ext = "json"
+            with requests.get(url, stream=True, timeout=10) as resp:
+                resp.raise_for_status()
+                with open("./out/template.json", "wb") as json_file:
+                    for chunk in resp.iter_content(chunk_size=8192):
+                        json_file.write(chunk)
 
-        with open(f"{document.slug}.csv") as csv_file:
-            self.upload_file(csv_file)
+    def main(self):
+        """If a template is provided, it extracts dataframes from the PDF(s)
+        using that template and appends it to a CSV for each document and returns
+        a zip file of all the CSVs. If no template is provided, it guesses
+        the boundaries for each file"""
+        with zipfile.ZipFile("export.zip", mode="w") as archive:
+            for document in self.get_documents():
+                url = self.data["url"]
+                if url is not None: #the if branch gets executed if a template is provided
+                    self.fetch_files(url)
```

Comment:
The URL is the same for all files, no need to re-download for each file.  Hoist this call out of the loop.